src/main/java/net/snowflake/ingest/streaming/internal/DataValidationUtil.java

```diff
@@ -1055,6 +1055,82 @@ static int validateAndParseBoolean(String columnName, Object input, long insertR
         insertRowIndex);
   }
 
+  /**
+   * Validate and cast Iceberg struct column to Map<String, Object>. Allowed Java type:
+   *
+   * <ul>
+   *   <li>Map<String, Object>
+   * </ul>
+   *
+   * @param columnName Column name, used in validation error messages
+   * @param input Object to validate and parse
+   * @param insertRowIndex Row index for error reporting
+   * @return Object cast to Map
+   */
+  static Map<String, ?> validateAndParseIcebergStruct(
+      String columnName, Object input, long insertRowIndex) {
+    if (!(input instanceof Map)) {
+      throw typeNotAllowedException(
+          columnName,
+          input.getClass(),
+          "STRUCT",
+          new String[] {"Map<String, Object>"},
+          insertRowIndex);
+    }
+    for (Object key : ((Map<?, ?>) input).keySet()) {
+      if (!(key instanceof String)) {
```

Comment:
we need to validate that the list of fields being passed in is a subset of the list of fields declared on the schema, no?

Reply:
Did it [here](https://github.com/snowflakedb/snowflake-ingest-java/blob/a1ab1074f657d8ed515a0b7fc90847f2f66c3955/src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java#L442). The `DataValidationUtil` only validates the data type, range, size etc. Schema related check are kept in `IcebergParquetValueParser`